File: src/email_automation_jev/repository.py
from __future__ import annotations

import re
from datetime import UTC, datetime

from supabase import Client, create_client

from .models import GmailAccount, NormalizedEmail, SyncCounts, SyncType


def utc_now() -> str:
    return datetime.now(UTC).isoformat().replace("+00:00", "Z")


def safe_error_message(error: BaseException) -> str:
    message = re.sub(r"Bearer\s+\S+", "Bearer [redacted]", str(error), flags=re.I)
    return message[:2000]


def create_database_client(url: str, service_role_key: str) -> Client:
    return create_client(url, service_role_key)


def ensure_gmail_account(database: Client, gmail_address: str) -> GmailAccount:
    response = database.table("gmail_accounts").upsert({"gmail_address": gmail_address.lower()}, on_conflict="gmail_address").execute()
    if not response.data:
        response = database.table("gmail_accounts").select("id,gmail_address,latest_history_id").eq("gmail_address", gmail_address.lower()).single().execute()
        row = response.data
    else:
        row = response.data[0]
    return GmailAccount(row["id"], row["gmail_address"], row.get("latest_history_id"))


def start_sync_run(database: Client, account_id: str, sync_type: SyncType) -> str:
    database.table("gmail_accounts").update({"sync_status": "running", "last_error": None}).eq("id", account_id).execute()
    response = database.table("sync_runs").insert({"gmail_account_id": account_id, "sync_type": sync_type}).execute()
    return str(response.data[0]["id"])


def upsert_emails(database: Client, account_id: str, emails: list[NormalizedEmail]) -> tuple[int, int]:
    if not emails:
        return 0, 0
    message_ids = [email["gmail_message_id"] for email in emails]
    existing_response = database.table("emails").select("gmail_message_id").eq("gmail_account_id", account_id).in_("gmail_message_id", message_ids).execute()
    existing = {row["gmail_message_id"] for row in existing_response.data}
    rows = [{"gmail_account_id": account_id, **email} for email in emails]
    database.table("emails").upsert(rows, on_conflict="gmail_account_id,gmail_message_id").execute()
    updated = sum(message_id in existing for message_id in message_ids)
    return len(message_ids) - updated, updated


def find_existing_message_ids(database: Client, account_id: str, ids: list[str]) -> set[str]:
    if not ids:
        return set()
    response = database.table("emails").select("gmail_message_id").eq("gmail_account_id", account_id).in_("gmail_message_id", ids).execute()
    return {row["gmail_message_id"] for row in response.data}


def mark_messages_deleted(database: Client, account_id: str, ids: list[str]) -> int:
    if not ids:
        return 0
    response = (
        database.table("emails")
        .update({"deleted_at": utc_now()})
        .eq("gmail_account_id", account_id)
        .in_("gmail_message_id", ids)
        .is_("deleted_at", "null")
        .execute()
    )
    return len(response.data)


def _count_values(counts: SyncCounts) -> dict[str, int]:
    return {
        "discovered_count": counts.discovered,
        "inserted_count": counts.inserted,
        "updated_count": counts.updated,
        "deleted_count": counts.deleted,
        "skipped_count": counts.skipped,
    }


def complete_sync_run(
    database: Client,
    *,
    account_id: str,
    run_id: str,
    history_id: str,
    counts: SyncCounts,
) -> None:
    finished_at = utc_now()
    database.table("sync_runs").update({"status": "succeeded", "finished_at": finished_at, **_count_values(counts)}).eq("id", run_id).execute()
    database.table("gmail_accounts").update(
        {
            "latest_history_id": history_id,
            "last_synced_at": finished_at,
            "sync_status": "idle",
            "last_error": None,
        }
    ).eq("id", account_id).execute()


def fail_sync_run(
    database: Client,
    *,
    account_id: str,
    run_id: str,
    error: BaseException,
    counts: SyncCounts,
) -> None:
    message = safe_error_message(error)
    database.table("sync_runs").update(
        {
            "status": "failed",
            "finished_at": utc_now(),
            "error_message": message,
            **_count_values(counts),
        }
    ).eq("id", run_id).execute()
    database.table("gmail_accounts").update({"sync_status": "error", "last_error": message}).eq("id", account_id).execute()

File: lib/repository.ts
import { createClient, type SupabaseClient } from "@supabase/supabase-js";
import type {
  GmailAccountRow,
  NormalizedEmail,
  SyncCounts,
  SyncType,
} from "./types.js";

type DatabaseClient = SupabaseClient;

function assertNoError(error: { message: string } | null, context: string): void {
  if (error) throw new Error(`${context}: ${error.message}`);
}

function safeErrorMessage(error: unknown): string {
  const message = error instanceof Error ? error.message : String(error);
  return message.replace(/Bearer\s+[^\s]+/gi, "Bearer [redacted]").slice(0, 2_000);
}

export function createDatabaseClient(url: string, serviceRoleKey: string): DatabaseClient {
  return createClient(url, serviceRoleKey, {
    auth: { persistSession: false, autoRefreshToken: false },
  });
}

export async function ensureGmailAccount(
  database: DatabaseClient,
  gmailAddress: string,
): Promise<GmailAccountRow> {
  const normalizedAddress = gmailAddress.toLowerCase();
  const { data, error } = await database
    .from("gmail_accounts")
    .upsert(
      { gmail_address: normalizedAddress },
      { onConflict: "gmail_address", ignoreDuplicates: false },
    )
    .select("id,gmail_address,latest_history_id")
    .single();
  assertNoError(error, "Could not create or load Gmail account");
  return data as GmailAccountRow;
}

export async function startSyncRun(
  database: DatabaseClient,
  accountId: string,
  syncType: SyncType,
): Promise<string> {
  const accountResult = await database
    .from("gmail_accounts")
    .update({ sync_status: "running", last_error: null })
    .eq("id", accountId);
  assertNoError(accountResult.error, "Could not mark account as syncing");

  const { data, error } = await database
    .from("sync_runs")
    .insert({ gmail_account_id: accountId, sync_type: syncType })
    .select("id")
    .single();
  assertNoError(error, "Could not create sync run");
  return (data as { id: string }).id;
}

export async function upsertEmails(
  database: DatabaseClient,
  accountId: string,
  emails: NormalizedEmail[],
): Promise<{ inserted: number; updated: number }> {
  if (emails.length === 0) return { inserted: 0, updated: 0 };

  const messageIds = emails.map((email) => email.gmail_message_id);
  const existingResult = await database
    .from("emails")
    .select("gmail_message_id")
    .eq("gmail_account_id", accountId)
    .in("gmail_message_id", messageIds);
  assertNoError(existingResult.error, "Could not inspect existing messages");
  const existing = new Set(
    (existingResult.data || []).map(
      (row) => (row as { gmail_message_id: string }).gmail_message_id,
    ),
  );

  const rows = emails.map((email) => ({
    gmail_account_id: accountId,
    ...email,
  }));
  const upsertResult = await database.from("emails").upsert(rows, {
    onConflict: "gmail_account_id,gmail_message_id",
    ignoreDuplicates: false,
  });
  assertNoError(upsertResult.error, "Could not upsert messages");

  const updated = messageIds.filter((id) => existing.has(id)).length;
  return { inserted: messageIds.length - updated, updated };
}

export async function findExistingMessageIds(
  database: DatabaseClient,
  accountId: string,
  messageIds: string[],
): Promise<Set<string>> {
  if (messageIds.length === 0) return new Set();
  const { data, error } = await database
    .from("emails")
    .select("gmail_message_id")
    .eq("gmail_account_id", accountId)
    .in("gmail_message_id", messageIds);
  assertNoError(error, "Could not inspect existing messages for resume");
  return new Set(
    (data || []).map(
      (row) => (row as { gmail_message_id: string }).gmail_message_id,
    ),
  );
}

export async function markMessagesDeleted(
  database: DatabaseClient,
  accountId: string,
  messageIds: string[],
): Promise<number> {
  if (messageIds.length === 0) return 0;
  const { data, error } = await database
    .from("emails")
    .update({ deleted_at: new Date().toISOString() })
    .eq("gmail_account_id", accountId)
    .in("gmail_message_id", messageIds)
    .is("deleted_at", null)
    .select("gmail_message_id");
  assertNoError(error, "Could not mark deleted messages");
  return data?.length || 0;
}

export async function completeSyncRun(
  database: DatabaseClient,
  options: {
    accountId: string;
    runId: string;
    historyId: string;
    counts: SyncCounts;
  },
): Promise<void> {
  const finishedAt = new Date().toISOString();
  const runResult = await database
    .from("sync_runs")
    .update({
      status: "succeeded",
      finished_at: finishedAt,
      discovered_count: options.counts.discovered,
      inserted_count: options.counts.inserted,
      updated_count: options.counts.updated,
      deleted_count: options.counts.deleted,
      skipped_count: options.counts.skipped,
    })
    .eq("id", options.runId);
  assertNoError(runResult.error, "Could not complete sync run");

  const accountResult = await database
    .from("gmail_accounts")
    .update({
      latest_history_id: options.historyId,
      last_synced_at: finishedAt,
      sync_status: "idle",
      last_error: null,
    })
    .eq("id", options.accountId);
  assertNoError(accountResult.error, "Could not advance Gmail history cursor");
}

export async function failSyncRun(
  database: DatabaseClient,
  options: { accountId: string; runId: string; error: unknown; counts: SyncCounts },
): Promise<void> {
  const errorMessage = safeErrorMessage(options.error);
  const finishedAt = new Date().toISOString();
  const [runResult, accountResult] = await Promise.all([
    database
      .from("sync_runs")
      .update({
        status: "failed",
        finished_at: finishedAt,
        discovered_count: options.counts.discovered,
        inserted_count: options.counts.inserted,
        updated_count: options.counts.updated,
        deleted_count: options.counts.deleted,
        skipped_count: options.counts.skipped,
        error_message: errorMessage,
      })
      .eq("id", options.runId),
    database
      .from("gmail_accounts")
      .update({ sync_status: "error", last_error: errorMessage })
      .eq("id", options.accountId),
  ]);
  assertNoError(runResult.error, "Could not record failed sync run");
  assertNoError(accountResult.error, "Could not mark Gmail account as failed");
}
